SpiffWorkflow/SpiffWorkflow/bpmn/specs/events/event_definitions.py

```diff
@@ -205,16 +206,18 @@ def update_task_data(self, my_task):
         if payload is not None:
             my_task.set_data(**payload)
 
-    def get_correlations(self, script_engine, payload):
+    def get_correlations(self, task, payload):
         correlations = {}
         for property in self.correlation_properties:
-            for key in property.correlation_keys:
-                if key not in correlations:
-                    correlations[key] = {}
-                correlations[key][property.name] = script_engine._evaluate(property.expression, payload)
```

Comment:
Will retain a get_correlation_keys method - and understand we may want to gather this information in this way at a future date.  However the structure is not helpful for determining if a message matches, or for storing the properties discovered through a send as properties can be shared between keys.

Reply:
It is absolutely helpful -- in fact it's all you need to know outside of SpiffWorkflow.  The correlation key (name and value) is the unit of comparison to use at the interprocess level, and in the code you're proposing to replace, Spiff is doing the work of collecting the all the properties into a unit, associating it with the name, and passing both along.

This change actually doesn't make too much difference internally to SpiffWorkflow, though it's conceptually cleaner to be comparing keys rather than individual properties even internally (that's the unit of comparison).  It does make a difference in applications relying on Spiff, which is why I think it's questionable, because it makes life in spiffworkflow-backend harder.

What ought to be changed about SpiffWorkflow is what information the parser returns about messages and correlations -- if the parser returned information about the correlation keys rather than their properties, you could rely on that and let SpiffWorkflow just provide you with a name and value which you can compare.  Instead, you're removing this from SpiffWorkflow (which can handle this easily) and replacing it with more complicated code in spiffworkflow-backend.

Reply:
So we make this call and grab the keys instead of the properties. It sees to me that in nearly all cases we would get back an array with a single topic/converation/key.  So let's just consider that simple case for now.  The respose from this function call might look something like this: ["invoice": {"customer_name":"Sartography", "PO":12}]

From this we would create a  Message Record in the backend DB that looks roughly like this:

RECEIVE MESSAGE:
type: RECEIVE
name:  'Request Approval'
correlation_key:   ["invoice": {"customer_name":"Sartography", "PO":12}]

When A SENDING Message comes in, it will contain a message name, and a payload.  We can't create a correlation key for the send message because it the receiving Message Task or Event that defines the extraction rules that would create correlation values.

SEND MESSAGE:
type: SEND
name:  'Request Approval'
payload:    {"cust_name":"Sartography", "purchase_order":12}]

We need to match these two messages outside of SpiffWorkflow Library because we can't instantiate / deserialize all workflow instances to determine if we have a match - it is reasonable to imagine a world where there are 100s or 1000s of waiting instances.   We also need to handle API Calls that might produce messages from external systems.  So we can't depend on having any more information about a SEND message than what is shown above.

CREATING A MATCH:
We take the message payload, and use it as the context for running each of the receive messages' correlation properties extraction rules against it.  These extraction rules are specific to a Message Instance -- they are not tied to the correlation keys, or the correlation properties, they are directly associated with the receiving message task or event.

So we have to loop through the currelation _properties_, run the eval, and get a set of values.
We can then build a dictionary that we could compare with the correlation key of the recieve message to determine if we have a match.  Because we can now store the correlation key as a json structure on the message object (and not a single value in a string filed in the database), we don't have the weird issue where we are converting things to strings for comparison.  

I think this method would work, and would raise the correlation_keys up to the right level in the backend.   So I'll give this a shot now.

Reply:
There actually doesn't need to be any calculation based on the message for catch events.  I know I said this the other day but I was wrong and should have looked at what the code I wrote actually did before making an assertion about it.

Sent messages DO contain correlation keys.  In fact, the correlation key is entirely determined by the sender (as the initiator of the conversation, they get to set the initial value).  When a message is sent, the correlation key and value is stored on the workflow.  Now this workflow can only receive messages that match the original key.

On the receiver workflow, at this point no correlations have been set, so it can accept the first matching message name.  Generally, the first message received would be a start event, so you start the process and send it the message.  The received then sets the correlation key and value on it's workflow, and at that point, it cannot accept messages that don't match the original key.

It is the message sender that does the calculations and it all happens inside Spiff.

What I'm proposing is a parse time table populated with:
process_model, message_name, start_event, throw_or_catch, correlation_key_name

You also have an instance table that looks something like this for waiting or future events to be caught:
process_instance, message_name, correlation_key_name, correlation_key_value

Without the proposed change, SpiffWorkflow sends you the message name, correlation key name, and the key value; the backend doesn't have to calculate anything.

When the backend receives a message, you can check your instance table for events that match all three returned values.

You can populate this table based on your parse time info.  When you receive a message that has no entries in your instance table, you go to the parse table and do a query for events that catch that message, and start any processes needed.  Then you insert rows for any other catch events in the original and started processes in the instance table along with the value of the correlation key sent by SpiffWorkflow.